lib/List/Util.pm

```diff
@@ -494,6 +495,51 @@ Returns the values of the input in a random order
 
 This function is affected by the C<$RAND> variable.
 
+=cut
+
+=head2 sample
+
+    my @items = sample $count, @values
+
+I<Since version 1.54.>
+
+Randomly select the given number of elements from the input list. Any given
+position in the input list will be selected at most once.
+
+If there are fewer than C<$count> items in the list then the result will be
+truncated; effectively the function becomes identical to L</shuffle>.
+
+This function is affected by the C<$RAND> variable.
+
+=cut
+
+sub _ffrac
+{
+  my $n = shift;
+  $n -= int $n;
+  $n++ if $n < 0;
+  $n;
+}
+
+sub sample
+{
+  my $count = shift;
+
+  my @ret;
+  while( @_ and $count-- ) {
+    my $i = $List::Util::RAND ?
+      int( _ffrac( $List::Util::RAND->() ) * @_ ) :
+      int rand @_;
+    push @ret, $_[$i];
+
+    # delete it from the list
+    $_[$i] = $_[-1] if $i < $#_; # No point doing this if $i is the final item
```

Comment:
This modifies the values of the array passed in. `sample 1, @foo` will change `@foo`

Reply:
Ughyes, you're right. A fix for that which still retains any kind of performance will be nontrivial. I guess a rewrite into XS is called for.

Now done at 624a150